Lya multiple scattering

jordanflitter opened the pull request:
This is a draft for the Lya multiple scattering branch.
There are now two new flags, called `LYA_MULTIPLE_SCATTERING` and `TEST_SL_WITH_MS_FILTER`, the latter becomes relevant only when the former is True, while the former is relevant only if `USE_HALO_FIELD=True`.
When `LYA_MULTIPLE_SCATTERING=True`, if `TEST_SL_WITH_MS_FILTER=False`, the code uses a modified window function that replaces the window function of Eq. (22) in [Davies, Mesinger, Murray 2025](https://arxiv.org/pdf/2504.17254) (it replaces only the window function that is applied on the SFRD, not the one that is applied on X-ray luminosity). It is a subtraction of two hypergeometric functions 2F3 (instead of a subtraction of two Bessel functions) where the parameters of 2F3, $\alpha$ and $\beta$, are determined by the current redshift $z$, the filter radius $R$ and the ionization fraction $x_\mathrm{HI}$. Unfortunately GSL doesn't support computing 2F3 so I had to come up with my own implementation, which works pretty well based on the tests that I made.
When `LYA_MULTIPLE_SCATTERING=True`, and `TEST_SL_WITH_MS_FILTER=True`, the code uses the same modified window function but with $\alpha\gg\beta$, as in this limit 2F3 approaches the Bessel function. This configuration should be used only as a sanity check that the new window function (multiple scattering = MS) yields the same output as the old one (straight line = SL) in the appropriate limit.
Currently, this branch is designed to work only when `USE_MINI_HALOS=False`. With mini-halos, the SFRD is also used for the Lyman-Werner feedback, but the applied filter should correspond to the old one (as LW photons travel in straight lines), which is not supported by the current implementation when `LYA_MULTIPLE_SCATTERING=True`. In the future, two new more fields in `XraySourceBox` (like `filtered_sfr_lw` and `filtered_sfr_lw_mini`) should be introduced in this setting (thereby increasing the amount of used memory), or alternatively we can throw an error when `USE_MINI_HALOS` and `LYA_MULTIPLE_SCATTERING` are both True.

Commit:
added a new multiple scattering window function

Commit:
added multiple scattering window function

Commit:
Included J_alpha as a new input

Commit:
Merge branch 'main' into mult_scatter

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jordanflitter committed:
added the option to have multiple scattering with mini-halos (required to introduce two new 4D arrays for LW flux as LW photons travel in straight lines). Also included USE_ADIABATIC_FLUCTUATIONS to allow the possibility to begin the simulation with homogeneous T_k

jordanflitter committed:
merged main

jordanflitter committed:
merged with main

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jordanflitter committed:
fixed bug and forced purging XraySource boxes to remove warning

jordanflitter committed:
Merge branch 'mult_scatter' of github.com:21cmfast/21cmfast into mult_scatter

jordanflitter committed:
Merge branch 'main' into mult_scatter

jordanflitter committed:
merged main into branch

jordanflitter committed:
merged main

jordanflitter committed:
updated mu and eta fit due to upgrade in the MCMC simulation

jordanflitter committed:
extrapolated fit to low values of x_em

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jordanflitter committed:
Merge branch 'main' into mult_scatter

jordanflitter committed:
merged main

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jordanflitter committed:
undo fix, see final solution in PR 554

jordanflitter committed:
Merge branch 'mult_scatter' of github.com:21cmfast/21cmfast into mult_scatter

jordanflitter committed:
Merge branch 'main' into mult_scatter

jordanflitter committed:
added heating/cooling rates to outputs

jordanflitter committed:
Fixed tests and removed the flag TEST_SL_WITH_MS_FILTER

jordanflitter committed:
Fixed tests

jordanflitter committed:
added more tests to the multiple scattering filter

jordanflitter committed:
Merge branch 'main' into mult_scatter

jordanflitter committed:
Matched notation with paper and referred to equations in the paper

jordanflitter committed:
added an error and tests with LYA_MULTIPLE_SCATTERING=True

jordanflitter committed:
Merge branch 'main' into mult_scatter

jordanflitter committed:
increased coverage

steven-murray reviewed (commented):
Thanks @jordanflitter as far as I can tell, this is well written and makes sense. I had only a few comments. In addition, it would be great to have a quick tutorial on this model (I know there's the paper, but having something very brief with plots would be good).

jordanflitter committed:
merged main

jordanflitter committed:
fixed a bug during merger (all tests pass now)

jordanflitter committed:
addressed Steven's comments

jordanflitter commented on src/py21cmfast/drivers/single_field.py:
```diff
@@ -461,6 +466,38 @@ def compute_xray_source_field(
+        # This approximation breaks when reionization begins, but under strong X-ray heating the value of the spin temperature (and
+        # Lyman alpha flux) becomes irrelevant. This should be examined in the future in the case of weak X-ray heating during the
+        # epoch of reionization (this would complicate things dramtically because a spatially varying filter implies
+        # that the box cannot be filtered in Fourier space via the convolution theorem!)
```
Done. I made a new issue #606.

jordanflitter commented on src/py21cmfast/src/filtering.c:
```diff
@@ -115,18 +116,204 @@ double spherical_shell_filter(double k, double R_outer, double R_inner) {
+    }
+}
+
+void initialize_alphas_and_betas_for_multiple_scattering(
```
Thank you very much for this comment! I agree that the implementation that you are suggesting is more elegant and less confusing. It also allowed me to update `test_hyper_2F3` and test it in more reasonable values of `alpha` and `beta`.

jordanflitter committed:
Merge branch 'main' into mult_scatter

jordanflitter committed:
removed arrays (that were helpful for the project) in order to reduce peak and storage memory. They will return

jordanflitter committed:
included a new models.rst in the docs and gave some explanation on the Lya multiple scattering effect. Also changed the maxlimit for large files because the figures take ~1.5MB

jordanflitter committed:
added more references to acknowledge.rst and show_references

jordanflitter commented:
Update of what was done since last review:

- The new arrays (that were useful for my project) were removed from this branch. They'll return after we fix #604.
- Included a preliminary `models.rst` in the docs. At the moment, I only gave information on the new flag, `LYA_MULTIPLE_SCATTERING`.
- Added my paper (and other papers that added new features to the code) to `acknowledge.rst` + `show_references`. Note that one of the papers that I added was Brad's paper about lightcone + RSD, I think it's an important paper that deserves some credit since before it came out there was no lightcone in 21cmFAST! Since `show_references` only accepted `inputs`, it couldn't know if to cite Brad's paper or not, so I added an argument called `lightcone` with default `True`.

jordanflitter committed:
increased coverage

Commit:
docs: improve input parameter docs

Commit:
fix: set all cstruct -> _cstruct

Commit:
fix: some incorrect attribute naming

Commit:
docs: typo in version number

steven-murray committed:
Merge pull request #616 from 21cmfast/doc-updates-mult-scatter

docs: improve input parameter docs

jordanflitter committed:
Merge branch 'main' into mult_scatter

jordanflitter committed:
Merge branch 'main' into mult_scatter

steven-murray committed:
Merge branch 'main' into mult_scatter

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci
